keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,854 @@
+- I can create a REST API (with TypeScript) for `downscalePodPicker` that runs in a Deployment, and answers a request to choose `N` Pods by:
+   1. keeping track of which region each Pod is in
+   2. finding `N` Pods that we can remove without violating the regional distribution requirements:
+       - if there are multiple acceptable options, we could choose the `N` Pods which are doing the least work
+   1. return these Pods as the "chosen pods"
+
+### Notes/Constraints/Caveats (Optional)
+
+- The `downscalePodPicker` APIs are written by users to suite their specific needs, and are not part of Kubernetes.
+- There are very few constraints on what a `downscalePodPicker` APIs can be, for example:
+   - they may be fully external to the Kubernetes cluster (e.g. running on an external metrics system)
+   - they may be part of the application which it controls (e.g. a REST API exposed by the application itself)
+   - they may be an adjacent Deployment running on the same Kubernetes cluster (e.g. a Python flask app)
+
+### Risks and Mitigations
+
+- A slow `downscalePodPicker` API may increase the time to downscale a ReplicaSet
+   - MITIGATION: we will provide a `timeoutSeconds` config, if the API has not successfully returned within this time, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- The `downscalePodPicker` may not be reachable at all times, or may return an invalid response
+   - MITIGATION: we will provide a `maxRetries` config, if the API continues to fail after this many attempts, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- Users may attempt to use `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` at the same time, resulting in unexpected behaviour
+   - MITIGATION: documentation will stress that only ONE of `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` should be used per ReplicaSet
+- A malicious (or poorly designed) `downscalePodPicker` API may attempt to overwhelm the controller by returning an extremely large payload
+   - MITIGATION: we will limit the maximum response size to something reasonable, like 1MB
+- Users may find it difficult to build their own `downscalePodPicker` API
+   - MITIGATION: documentation will provide example implementations in common programing languages
+- Users may create non-idempotent `downscalePodPicker` APIs that have side effects (e.g. preventing chosen workers from accepting new tasks):
+   - MITIGATION: documentation will stress that `downscalePodPicker` API must be idempotent, and may be called multiple times
+- Malicious actors may call `downscalePodPicker` APIs from outside the controller loop
+   - MITIGATION: documentation will stress that `downscalePodPicker` APIs should be secured by authentication at the network or request layer
+
+
+## Design Details
+
+#### API Contract:
```

Comment:
external API contract.

We usually start KEPs with the k8s API spec changes.

Reply:
I have renamed this to `"Pod Picker" API Contract` in https://github.com/kubernetes/enhancements/commit/8beae1b5f2f9e7ed7814ce7a711c8633fe6f46d3